### DISCUSSION TITLE
Using PIXI JS with THREE

### DISCUSSION THREAD

**Discussion by StrangeScript** [pixijs/pixijs#7767] [Q&A] [answered]
Hello everyone,
I am making a game in threejs, and I want to use pixi for the user interface and the menus.
Currently I import the domElement from the threejs renderer as a texture into pixijs. But this doesnt seem to be a very optimized way.
Would it be possible to make three js draw over the gl context and let the pixi renderer draw over it?
or could I import a threejs texture directly into pixi, without drawing to the domelement first?

If somebody has before combined these libraries, please share what you think is the best way
Thank you :)

**Comment by ivanpopelyshev** [comment, marked as answer]
Yes, I used three with pixi and others did it too, whats your problem?

just use same context & dont forget to

```
threeRenderer.reset();
threeRenderer.render()
pixiRenderer.reset();
pixiRenderer.render();
pixiRenderer.reset();
```

something like that. one or two resets.
However, while Pixi does actually treat `reset()` as `someone else used the context, empty the state`, threejs sometimes doesnt. In this case, if you are not familiar with webgl and what is state consists of - you'll have some problems :)

**Reply by StrangeScript** [reply]
Thank you for answer, god of pixi,

The problem for me is that pixi overwrites the whole scene.
I am probably doing it wrong, as I am indeed not yet too familiar with everything of webgl.

I made a small jsfiddle with pixi and three, at the bottom I put the render function, but its not working.
https://jsfiddle.net/r8mndupL/

If both renderers use the same canvas element, do they automatically use the same context?

**Reply by ivanpopelyshev** [reply]
https://jsfiddle.net/Hackerham/2uLkjpa8/3/

1. threejs method that resets state: https://github.com/mrdoob/three.js/blob/a2f66aff6a7f2b819bdb882021404c92b7a1ea0a/src/renderers/WebGLRenderer.js#L2222-L2231
2. pass webgl context taken by threejs to pixi renderer creation: https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L334-L338
3. pixi should not clear the framebuffer `render(stage, {clear:false})`

Did you clone threejs and pixi on your computer and opened it in separate IDE windows for better search?

**Reply by ivanpopelyshev** [reply]
Please go to https://github.com/pixijs/guides , and make an article there.

**Reply by StrangeScript** [reply]
Thank you so much Ivan, that was very helpfull. I understand it now

**Reply by miltoncandelero** [reply]
> Please go to https://github.com/pixijs/guides , and make an article there.

I will at least make some examples for the pixi/examples 👍

**Reply by ivanpopelyshev** [reply]
Adding link to threejs build in pixi examples should be fun :)

**Comment by humbleCoder1990** [comment]
@ivanpopelyshev 

I would like to know why using pixi & three both.
Is there something pixi can't do while three can?

**Reply by StrangeScript** [reply]
@humbleCoder1990 
I am using three js because its a 3d library, and PIXI for 2d user interface graphics.

**Comment by eviltik** [comment]
Hi there and thank you for this discussion, it help.

I'm testing pixijs + threejs + webxr. It have some interesting results. But required some hacks.

### pixijs v6
Because i plan to crosspost in threejs community, just a little intro about pixijs.

Pixijs is a 2D javascript library known for its [speed](https://benchmarks.slaylines.io/). From what I understand, it tries to use webgl by default, then fallback to canvas if unavailable. I'm new to this library so I haven't looked at the approach used to achieve this good level of performance. Does canvas backend use workers ? probably yes. Is there any webassembly somewhere ? i doubt. WebGL magics ? yes ! Well coded ? YES. The API is very convenient.

### use case:

```
Use pixijs as a fast canvas backend generator in threejs, using a canvas texture. 
Use the interaction layer / events of pixijs to make a nice bind.
```

I use pixijs to generate animated interactive buttons in a threejs scene for now. The tricky part is the `interactive mode` (pixijs feature). When interactive mode is enabled on a displayObject, it add pointers and mouse events listeners on the window and on the canvas. Finaly you can  `displayObject.on('click', ...)`; so easy : pixijs take care of all for you (mouse and pointer events).

But in a threejs context, a 3D context, handling mouse and pointers events directly does not make sense. We have to deal with raycasters. Mouse raycasters, controllers (headset) raycasters. I didn't found a nice how to related to that. Does somebody already play with this use case in the room ?
 
### pixijs + threejs

The pixijs + threejs assembly takes the form of a `THREE.Mesh`, composed of a `THREE.PlaneGeometry` and a `THREE.CanvasTexture`.

I noticed theses points below and coded accordingly:
* in immersive mode, the event loop (tickers loop ?) of pixijs stops. I don't know why but probably for a good reason.
* rather than tickers loop, i used the threejs setAnimation loop for animations and `pixijs.renderer()` manually (works well)

I am currently trying to solve the following:
* it doesn't seem possible to keep a pixijs object "interactive" without pixijs directly bind the mouse, thus retrieving incorrect informations in a 3D context
* but if I don't enable the interactive mode, it will not handle Pointers or MouseEvents, so I can't simulate sending a click with correct coordinates 
* I have to deal with input raycasters and find a way to simulate mouse events to the pixijs library with correct coordinates (works well ..)

```
// pixiMesh is a threejs plane with pixi canvas texture

// below some code related to three.js

// define meshes allowed to be raycasted
const threeJsRaycasterObjects = [ pixiMesh ]

// when it make sense intersectObjects feature will provide a list of meshes
// that the raycaster hit. In our use case, it append when hovering the pixiMesh.
// mouse: mousemove, up, down
// webxr: selectstart, selectend, select
const intersectedMeshes = myRaycaster.intersectObjects( threeJsRaycasterObjects );
```

I can find X/Y coordinate, create a mouse event, then dispatch it on the pixi canvas,
simulating a mouse behavior.
```
    // create an mouse event
    const ev = new MouseEvent( 'mousedown', {
	    bubbles: true,
		cancelable: true,
		clientX: hoveredPixiMesh.uv.x * pixiCanvasWidth,
		clientY: hoveredPixiMesh.uv.y * pixiCanvasHeight
	} );

    // send it to pixijs trought an event on his canvas
    threejsCanvasTexture.image.dispatchEvent( ev );
```

At this point, i can simulate most of mouse and pointer events as a proxy between treejs and pixijs but i can't take keep benefits of the pixijs interactive mode, because some methods are private and i can't use them.

Purposal: 
* every onPointer* private methods should be public : we can programmatically trigger events while interactive mode is false
* or add some public method calling theses private methods

Methods list :
* onPointerMove
* onPointerUp
* onPointerDown
* onPointerOverOut
* onWheel

https://github.com/pixijs/pixijs/blob/dev/packages/events/src/EventSystem.ts

I didn't test this change, but perhaps somebody have a complete different way to handle this use case. Just asking before. Thank you !

![image](https://user-images.githubusercontent.com/4477890/163011922-c4e88c67-9be5-438b-9285-da514a4c9579.png)

**Reply by miltoncandelero** [reply]
Wait... if you are only using pixi to generate canvas textures... why would you use the pointer events? they make sense only when you paint to the screen as those events map a point on screen to a point in the renderer and thus, the stage.

**Reply by eviltik** [reply]
let's say i want to paint to the screen (animated button hover effect, blur, slide ..)

**Reply by eviltik** [reply]
wait, perhaps this is the way to go https://api.pixijs.io/@pixi/interaction/PIXI/InteractionManager.html

**Reply by eviltik** [reply]
> Wait... if you are only using pixi to generate canvas textures... why would you use the pointer events? they make sense only when you paint to the screen as those events map a point on screen to a point in the renderer and thus, the stage.

Let me add a point. In the screenshot, you can see a dark panel and a button. This is a unique pixi canvas texture instance, and i can have multiple buttons in a single texture rather than plenty of three meshes with it's own pixi instance.

**Reply by eviltik** [reply]
It sound's i have to create a custom InteractionManager for this context (threejs), right @miltoncandelero ?

**Reply by jonlepage** [reply]
yes , ex on my side, when i run my editor , is attache to pixijs.
and i just use a getter to hack the interaction of pixijs.
```ts
export class ApplicationEditor {

	/** getter and not create because they not crush,pixi emit x2  */
	get interaction():InteractionManager {
		return $app.renderer.plugins.interaction;
	}
}
```

interaction manager are not nicely documented, so you will need make many test to understand it😯
the  current mouse data a stored here.
![image](https://user-images.githubusercontent.com/24865815/163044557-732a9152-26a4-40f1-b7a4-a6bd3b00a4c4.png)


pixisj also use `eventemitter3` libs.
so you can also import it and create and bind your own immiter .
ex :
```ts
	subscribesEventEmitters(){
		$EE.on( EE.onSceneChange, this.onChaneScene, this );
	}
```

**Comment by goldenratio** [comment]
@StrangeScript have you looked into Pixi3D https://github.com/jnsmalm/pixi3d It has good support for glTF. It works well in our already existing 2D ecosystem.
